src/main/java/javaslang/collection/Iterator.java

```diff
@@ -166,6 +167,19 @@ public T next() {
         };
     }
 
+    @Override
+    Iterator<T> peek(Consumer<? super T> action);
```

Comment:
Because Interable is lazy, i.e. evaluates values only, when needed, when can take an approch similar to Stream.peek(Consumer):

```java
@Override
default Iterator<T> peek(Consumer<? super T> action) {
    Objects.requireNonNull(action, "action is null");
    if (!hasNext()) {
        return Iterator.empty();
    } else {
        final Iterator<T> that = this;
        return new Iterator<T>() {
            @Override
             public boolean hasNext() {
                 return that.hasNext();
             }

            @Override
             public T next() {
                 if (!hasNext()) {
                     throw new NoSuchElementException();
                 }
                 final T next = that.next();
                 consumer.accept(next);
                 return next;
             }
}
```

Please remove the abstract class `Impl`, we do not need it.

I hope you dont't mind to revert a part of your work. I really appreciate your help!

Greets,

Daniel

PS: I only have an iPad here. The code above is not tetsted, it may not work.